### ISSUE TITLE
objects returned by hb_directwrite_face_create aren't usable

### ISSUE THREAD

**Issue by matthiasclasen** [harfbuzz/harfbuzz#3683]
What I mean is that I have a hard time creating a cairo scaled font for such a hb_font_t.

We first tried to use cairo_dwrite_font_face_create_for_dwrite_fontface, but that fails because hb_directwrite_face_get_font_face doesn't return the object that the hb_font_t was created from.

Then we tried to use cairo-ft. That fails because hb_face_reference_blob doesn't give me the font data for such fonts.

The situation for coretext fonts on macOS may be a little better because hb_coretext_font_create seems to keep around the native object. Can the directwrite api be fixed to do the same?

**Comment by matthiasclasen** [comment]
Or maybe there is an alternative scenario where I should use hb_face_builder in pango to create hb_face_t objects that have a blob?

**Comment by behdad** [comment]
> The nicest (and for me, easiest) would be if we had
> 
> ```
> cairo_font_face_t *cairo_hb_font_face_create_for_hb_face (hb_font_face_t *face)
> ```
> 
> it looks like that would work nicely, by using hb_face_reference_table for the load_truetype_table implementation

Right. Or cairo had a load-table constructor for font-face...

**Comment by matthiasclasen** [comment]
https://gitlab.gnome.org/GNOME/pango/-/merge_requests/593/diffs?commit_id=6a21d3456ee61b0b1adb8f1ce8b30c3772fcc0d7

is the workaround I came up with, using the face builder

**Comment by behdad** [comment]
@matthiasclasen see if #3686 helps?

**Comment by behdad** [comment]
Another question is: in the `_get_` API: if the font was NOT created with a native object, but we already have created a native object internally, should we return that? That happens if user has shaped something with the respective native shaper already.

**Comment by behdad** [comment]
> Another question is: in the `_get_` API: if the font was NOT created with a native object, but we already have created a native object internally, should we return that? That happens if user has shaped something with the respective native shaper already.

I'm leaning toward no. But for CoreText shaper this becomes an issue the way we currently do things...

**Comment by matthiasclasen** [comment]
> I'm leaning toward no.

Yeah, separating that out might be cleaner. 

What will happen in cross-api scenarios ? Create a face with hb_coretext_face_create, then call hb_ft_font_lock_face

**Comment by behdad** [comment]
> What will happen in cross-api scenarios ? Create a face with hb_coretext_face_create, then call hb_ft_font_lock_face

Right now it returns null.

In the `or_create` version, which we won't add for `lock_face`, just for `get_face`, it will create one using get_table API.

**Comment by matthiasclasen** [comment]
Can you always create a native object matching an arbitrary hb_face_t?

**Comment by behdad** [comment]
> Can you always create a native object matching an arbitrary hb_face_t?

Yes, because hb-face always have a get-table callback, and all systems can create a native object using just that.

**Comment by matthiasclasen** [comment]
Another thing to consider while we are looking at this is change notification:

hb_ft has hb_ft_font_changed that you are meant to call when you change the font axes.

coretext and directwrite don't seem to have any provision for that.

And I wonder why _I_ have to call this api. Shouldn't calling hb_font_set_var_coords_design be enough and do the right thing?

**Comment by matthiasclasen** [comment]
How is change notification handled wrt to native objects?

I see font->serial, and I guess callers are expected to stash the serial and check it every now and then. But does harfbuzz clear font->data.coretext when the serial is bumped ?

**Comment by matthiasclasen** [comment]
The handling of variations with native-backed objects seems unclear at best.

After

```
face = hb_ft_face_create_referenced (ft_face);
font1 = hb_font_create (face);
font2 = hb_font_create (face);
```

all 3 objects will share the underlying FT_Face. And calling

```
hb_font_set_variations (font1, ....);
```
Or am I missing something?
will modify that underlying FT_Face.

**Comment by matthiasclasen** [comment]
Thinking about my use, I don't really care about 'get back the original' at all.

What I need is

```
hb_face_t *
hb_NATIVE_face_create (NATIVE *native)

NATIVE *
hb_NATIVE_font_get_NATIVE (hb_font_t *font)
```
where the constructor produces a hb object that I can use for creating fonts that can be used for shaping
and the getter returns a native object that can be used for rendering (ie it should have size and variations applied).

**Comment by matthiasclasen** [comment]
The part that I find slightly confusing is this:

If I create two fonts, once this way:

```
font1 = hb_ft_font_create (ft_face);
font2 = hb_ft_font_create (ft_face);
```
and then this way:
```
face = hb_ft_face_create (ft_face);
font1 = hb_font_create (face);
font2 = hb_font_create (face);
```

then the resulting objects behave differently.

**Comment by behdad** [comment]
> If I create two fonts, once this way:
> 
> ```
> font1 = hb_ft_font_create (ft_face);
> font2 = hb_ft_font_create (ft_face);
> ```
> 
> and then this way:
> 
> ```
> face = hb_ft_face_create (ft_face);
> font1 = hb_font_create (face);
> font2 = hb_font_create (face);
> ```
> 
> then the resulting objects behave differently.

Right. The latter creates two fonts in their default scale and variations. The former copies scale/variation from ft_face. That's implied in the dataflow...

**Comment by behdad** [comment]
> I see font->serial, and I guess callers are expected to stash the serial and check it every now and then. But does harfbuzz clear font->data.coretext when the serial is bumped ? I see such a serial check in hb-ft.cc, but nothing in hb-coretext.cc or hb-directwrite.cc

Coretext had some hack to do that. Directwrite doesn't need it because the shaper just uses the font scale at shape time and doesn't store any native data otherwise. (ie. there's no directwrite font we keep, just face). I just pushed a change to drop all font-related shaper data whenever the font changes.

**Comment by matthiasclasen** [comment]
> Right. The latter creates two fonts in their default scale and variations. The former copies scale/variation from ft_face. That's implied in the dataflow...

The thing is: in the latter case, the two fonts are connected at the hip, like Siamese twins, via the shared underlying native object.

Maybe this is just not a thing to worry too much about. I don't use hb-ft, after all, and all my hb objects are immutable...

**Comment by matthiasclasen** [comment]
See https://github.com/harfbuzz/harfbuzz/pull/3696

**Comment by behdad** [comment]
`hb-directwrite` has changed a lot now. Can you look into it again?

**Comment by khaledhosny** [comment]
I suppose this is working now, please re-open if it isn’t.
